vars/pythonSetupPyPipeline.groovy

```diff
@@ -49,96 +61,212 @@ def call(Map pipelineParams) {
+              skipDefaultCheckout()
+            }
+            when {
+              beforeAgent true
+              allOf {
+                branch "master"
+                expression { params.doRelease }
+              }
+            }
+            steps {
+              unstash "pypi"
+              withCredentials([usernamePassword(credentialsId: pipelineParams.pypiCredentials, usernameVariable: "USERNAME", passwordVariable: "PASSWORD")]) {
+                sh "twine upload --verbose -u $USERNAME -p $PASSWORD --repository-url ${pipelineParams.pypiRepo} dist/*"
+              }
+            }
+          } // Deploy PyPI
+
+          stage("Deploy Docker") {
+            agent {
+              node {
+                label "docker"
+              }
+            }
+            options {
+              // No need to checkout sources again since we use stashed artifacts for deployment
+              skipDefaultCheckout()
+            }
+            when {
+              beforeAgent true
+              expression { pipelineParams.dockerDeploy }
+            }
+            steps {
+              unstash "docker"
+              unstash "pypi"
+              script {
+                docker.withRegistry(pipelineParams.dockerRegistryUrl, pipelineParams.dockerRegistryCredentialsId) {
+                  def image = docker.image("${pipelineParams.dockerRepo}/${moduleName}:${moduleVersion}")
+                  def imageExists = true
+                  try {
+                    image.pull()
+                  } catch(Exception) {
+                    echo "Image '${image.imageName()}' already exists in registry"
+                    imageExists = false
+                  }
+
+                  // Snapshot images can be overwritten, whereas a release one shouldn't be.
+                  if (isSnapshot(moduleVersion) || !imageExists) {
+                    // Until https://github.com/jenkinsci/docker-workflow-plugin/pull/162 is merged we have to
+                    // build and push the image directly via docker commands.
+                    sh "docker build -t ${image.imageName()} -f ${pipelineParams.dockerDeployFile} ${pipelineParams.dockerBuildArgs} ."
+                    sh "docker push ${image.imageName()}"
+                  }
+                }
+              }
+            }
+          } // Deploy Docker
         }
-        steps {
-          withCredentials([usernamePassword(credentialsId: pipelineParams.pypiCredentials, usernameVariable: "USERNAME", passwordVariable: "PASSWORD")]) {
-            sh "twine upload --verbose -u $USERNAME -p $PASSWORD --repository-url ${pipelineParams.pypiRepo} dist/*"
+      } // Deploy
+
+      stage("Bump version Patch") {
+        agent {
+          dockerfile {
+            filename pipelineParams.dockerFilename
+            additionalBuildArgs pipelineParams.dockerBuildArgs
+            args pipelineParams.dockerRunArgs
+            reuseNode true
```

Comment:
what we get from reuseNode here?

Reply:
Based on the current status this part is needed to ensure that we run always on the same node. Why? I am doing a thorough cleanup on the node and wiping the workspace + the created docker image and all the intermediate ones that Jenkins does not delete